rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/130_knn_query_nested_search.yml

```diff
@@ -0,0 +1,212 @@
+            nested:
+              path: nested
+              query:
+                knn:
+                  field: nested.vector
+                  query_vector: [ -0.5, 90.0, -10, 14.8, -156.0 ]
+                  num_candidates: 1
+              inner_hits: { size: 1, "fields": [ "nested.paragraph_id" ], _source: false }
+
+  - match: {hits.total.value: 1}  # as alias is passed as pre-filter, we get a single result
+  - match: {hits.hits.0._id: "3"}
+  - match: {hits.hits.0.fields.name.0: "rabbit.jpg"}
+  - match: { hits.hits.0.inner_hits.nested.hits.hits.0.fields.nested.0.paragraph_id.0: "0" }
+
+---
+"nested kNN search post-filtered on top level fields":
+  - do:
+      search:
+        index: test
+        body:
+          fields: [ "name" ]
+          query:
+            bool:
+              must:
+                - term:
+                    name: "rabbit.jpg"
+                - nested:
+                    path: nested
+                    query:
+                      knn:
+                        field: nested.vector
+                        query_vector: [ -0.5, 90.0, -10, 14.8, -156.0 ]
+                        num_candidates: 1
+  - match: { hits.total.value: 0 } # no hits because returned single vector did not pass post-filter
+
+  - do:
+      search:
+        index: test
+        body:
+          fields: [ "name" ]
+          query:
+            bool:
+              must:
+                - term:
+                    name: "rabbit.jpg"
+                - nested:
+                    path: nested
+                    query:
+                      knn:
+                        field: nested.vector
+                        query_vector: [ -0.5, 90.0, -10, 14.8, -156.0 ]
+                        num_candidates: 3
+                    inner_hits: { size: 1, fields: [ "nested.paragraph_id" ], _source: false }
+
+  - match: {hits.total.value: 1}
+  - match: {hits.hits.0._id: "3"}
+  - match: {hits.hits.0.fields.name.0: "rabbit.jpg"}
+  - match: { hits.hits.0.inner_hits.nested.hits.hits.0.fields.nested.0.paragraph_id.0: "0" }
+---
+
+"nested kNN search post-filtered on nested fields DOES NOT work":
+  - do:
+      search:
+        index: test
+        body:
+          fields: [ "name" ]
+          query:
+            nested:
+              path: nested
+              query:
+                bool:
+                  must:
+                    - term:
+                        nested.paragraph_id: 3
+                    - knn:
+                        field: nested.vector
+                        query_vector: [ -0.5, 90.0, -10, 14.8, -156.0 ]
+                        num_candidates: 6
+              inner_hits: { size: 1, "fields": [ "nested.paragraph_id" ], _source: false }
+  # no hits because, regardless of num_candidates knn returns top 3 child vectors from distinct parents
+  # and they don't pass the post-filter
+  # TODO: fix it on Lucene leve so nested knn respects num_candidates
+  - match: {hits.total.value: 0}
```

Comment:
This is indeed tricky. It does satisfy `num_candidates`, but how we are pre-filtering things can act weird with other nested queries. You are correct, it is worth revisiting in the future.

Along with allowing nested pre-filtering (filtering on child field values instead of just parent values).